src/ninja.cc

```diff
@@ -505,6 +507,213 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
+      for (int i = 0; i < dep_nodes_count; ++i) {
+        if (dep_nodes[i]->in_edge() == *ne) {
+          generated_nodes_.insert(dep_nodes[i]);
+          printf("Missing dep: %s uses %s (generated by %s)\n",
+              node->path().c_str(), dep_nodes[i]->path().c_str(),
+              (*ne)->rule().name().c_str());
+          generator_rules_.insert(&(*ne)->rule());
+          missing_deps_rule_names.insert((*ne)->rule().name());
+        }
+      }
+    }
+    missing_dep_path_count_ += missing_deps_rule_names.size();
+    nodes_missing_deps_.insert(node);
+  }
+}
+
+int MissingDependencyScanner::PrintSummary() {
+  printf("Processed %d nodes.\n", (int)seen_.size());
+  int retval = 0;
+  if (!nodes_missing_gen_.empty()) {
+    printf(
+        "Error: %d file(s) under %s are deps log dependencies of other "
+        "targets, but have no generator target listing them as outputs.\n",
+        (int)nodes_missing_gen_.size(), abs_gen_dir_.c_str());
+    retval += 1;
+  }
+  if (!nodes_missing_deps_.empty()) {
+    printf("Error: There are %d missing dependency paths.\n",
+           missing_dep_path_count_);
+    printf(
+        "%d targets had depfile dependencies on %d distinct generated inputs "
+        "(from %d rules) without a non-depfile dep path to the generator.\n",
+        (int)nodes_missing_deps_.size(), (int)generated_nodes_.size(),
+        (int)generator_rules_.size());
+    retval += 2;
+  }
+  if (retval) {
+    printf(
+        "There might be build flakiness if any of the targets listed above are "
+        "built alone, or not late enough, in a clean output directory.\n");
+  } else {
+    printf("No missing dependencies on generated files found.\n");
+  }
+  return retval;
+}
+
+int NinjaMain::ToolMissingDeps(const Options* options, int argc, char** argv) {
+  // This tool uses getopt, and expects argv[0] to contain the name of
+  // the tool, i.e. "missingdeps".
+  argc++;
+  argv--;
+
+  string gen_dir = "";
+
+  optind = 1;
+  int opt;
+  while ((opt = getopt(argc, argv, const_cast<char*>("g:h"))) != -1) {
+    switch (opt) {
+    case 'g':
+      gen_dir = optarg;
+      break;
+    case 'h':
+    default:
+      printf("usage: ninja -t missingdeps [options] [targets]\n"
+"\n"
+"options:\n"
+"  -g DIR   assume files under DIR are generated (default: .)\n"
+"    DIR is relative to the build directory, empty string disables\n"
+"    checking for files with a missing generator\n"
+             );
+      return 1;
+    }
+  }
+  argv += optind;
+  argc -= optind;
+
+  vector<Node*> nodes;
+  string err;
+  if (!CollectTargetsFromArgs(argc, argv, &nodes, &err)) {
+    Error("%s", err.c_str());
+    return 1;
+  }
+  RealDiskInterface disk_interface;
+  MissingDependencyScanner scanner(&deps_log_, &state_, &disk_interface,
+                                   gen_dir);
+  for (vector<Node*>::iterator it = nodes.begin(); it != nodes.end(); ++it) {
+    scanner.ProcessNode(*it);
+  }
```

Comment:
Can we have gtest base test for MissingDependencyScanner?

Reply:
OK, I will have to see what I can come up with. Are there any existing tests that set up a build graph and depfile?

Reply:
gtests now included